async_grpc/async_client.h

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CPP_GRPC_ASYNC_CLIENT_H
+#define CPP_GRPC_ASYNC_CLIENT_H
+
+#include <memory>
+
+#include "async_grpc/rpc_service_method_traits.h"
+#include "common/make_unique.h"
+#include "completion_queue_pool.h"
+#include "glog/logging.h"
+#include "grpc++/grpc++.h"
+#include "grpc++/impl/codegen/async_stream.h"
+#include "grpc++/impl/codegen/async_unary_call.h"
+#include "grpc++/impl/codegen/proto_utils.h"
+
+namespace async_grpc {
+
+class ClientEvent;
+
+class AsyncClientInterface {
+ public:
+  virtual ~AsyncClientInterface() = default;
+
+  virtual void HandleEvent(const ClientEvent& client_event) = 0;
+};
+
+template <typename RpcServiceMethodConcept,
+          ::grpc::internal::RpcMethod::RpcType StreamType =
+              RpcServiceMethodTraits<RpcServiceMethodConcept>::StreamType>
+class AsyncClient : public AsyncClientInterface {};
```

Comment:
I think the typing can be stricter because the caller does not need the public handleevent function.
Is it possible to 
- remove the inheritance to the interface
- introduce a nested class that has the handleevent function and owns the response_reader_, response_ and callback_?

Reply:
done